### DISCUSSION TITLE
`@mantine/form`: how to subscribe to form's touched and dirty states?

### DISCUSSION THREAD

**Discussion by park-jemin** [mantinedev/mantine#6406] [Q&A]
I would like to be able to disable my submit button until my form is dirty - what would be the way to do that using `@mantine/form`?

It would be nice to have a way to subscribe to form state, like how `react-hook-form` has [`useFormState`](https://react-hook-form.com/docs/useformstate) and other exposed APIs around form state.

**Comment by rtivital** [comment]
Use `{ mode: 'controlled' }`, it updates all values real time

**Reply by park-jemin** [reply]
Is there a way to do this _without_ setting the form to controlled mode? I would prefer to keep my form in uncontrolled mode to because it's a pretty large form. At the moment, I have something like this to mimic what I need, but it feels like something like this should be exposed via the library:

```ts
import { useState } from 'react'

import type { UseFormReturnType } from '@mantine/form'

export const useFormState = <TValues extends Record<string, unknown>>(
  form: UseFormReturnType<TValues>
) => {
  const [isDirty, setIsDirty] = useState(false)
  const [isTouched, setIsTouched] = useState(false)

  Object.keys(form.values).forEach(key =>
    form.watch(key, () => {
      setIsDirty(form.isDirty())
      setIsTouched(form.isTouched())
    })
  )

  return { isDirty, isTouched }
}

```

**Reply by rtivital** [reply]
No, currently, you can get these values only in controlled mode

**Reply by albert-sway** [reply]
Just throwing it out there as a data point that this is a pretty important thing for us (to be able to use the dirty state in an uncontrolled form) -- we have a large form and we don't want the save button enabled if the form isn't dirty. At the moment, we just can't implement that feature (the workaround above didn't work for us for some reason).

It would be awesome, from our perspective, if there were hooks we could use to subscribe to various form changes (like `useFormEffect`).
